File: scripts/inventory/ItemRegistry.cs
using System.Collections.Generic;
using System.Text.Json;
using Godot;

namespace ProiectDeLegume.Scripts.Inventory;

public static class ItemRegistry
{
    private static readonly Dictionary<string, ItemDef> Items = new();

    public static void Load()
    {
        var file = FileAccess.Open("res://assets/data/items.json", FileAccess.ModeFlags.Read);
        if (file == null)
        {
            GD.PrintErr("Cannot open items.json");
            return;
        }

        string jsonText = file.GetAsText();
        file.Close();

        var itemList = JsonSerializer.Deserialize<List<ItemDef>>(jsonText);
        if (itemList == null) return;

        foreach (var item in itemList)
        {
            Items[item.Id] = item;
        }

        GD.Print($"Loaded {Items.Count} items");
    }

    public static ItemDef Get(string id)
    {
        return Items.TryGetValue(id, out var item) ? item : throw new KeyNotFoundException($"Unknown item: {id}");
    }

    public static IEnumerable<ItemDef> All() => Items.Values;
}

File: src/commonMain/kotlin/inventory/ItemRegistry.kt
package inventory

import korlibs.io.file.std.*
import kotlinx.serialization.json.Json

object ItemRegistry {
    private val items = mutableMapOf<String, ItemDef>()

    suspend fun load() {
        val jsonText = resourcesVfs["data/items.json"].readString()
        val itemList = Json.decodeFromString<List<ItemDef>>(jsonText)
        for (item in itemList) {
            items[item.id] = item
        }
    }

    fun get(id: String): ItemDef = items[id] ?: error("Unknown item: $id")

    fun all(): Collection<ItemDef> = items.values
}
